dbms/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp

```diff
@@ -134,52 +133,72 @@ void ReadBufferFromKafkaConsumer::unsubscribe()
     consumer->unsubscribe();
 }
 
-/// Do commit messages implicitly after we processed the previous batch.
-bool ReadBufferFromKafkaConsumer::nextImpl()
+bool ReadBufferFromKafkaConsumer::reset()
 {
-    /// NOTE: ReadBuffer was implemented with an immutable underlying contents in mind.
-    ///       If we failed to poll any message once - don't try again.
-    ///       Otherwise, the |poll_timeout| expectations get flawn.
-    if (stalled || stopped || !allowed)
-        return false;
-
-    if (current == messages.end())
+    if (current != messages.end())
     {
-        if (intermediate_commit)
-            commit();
+        ++current;
+        return true;
+    }
+
+    return poll(0);
+}
 
-        /// Don't drop old messages immediately, since we may need them for virtual columns.
-        auto new_messages = consumer->poll_batch(batch_size, std::chrono::milliseconds(poll_timeout));
-        if (new_messages.empty())
+bool ReadBufferFromKafkaConsumer::nextImpl()
+{
+    while (current != messages.end())
+    {
+        if (auto error = current->get_error())
         {
-            LOG_TRACE(log, "Stalled");
-            stalled = true;
-            return false;
+            /// FIXME: should throw exception instead
+            LOG_ERROR(log, "Consumer error: " << error);
+            ++current;
+            continue;
         }
-        messages = std::move(new_messages);
-        current = messages.begin();
 
-        LOG_TRACE(log, "Polled batch of " << messages.size() << " messages");
+        break;
     }
 
-    if (auto err = current->get_error())
-    {
-        ++current;
-
-        // TODO: should throw exception instead
-        LOG_ERROR(log, "Consumer error: " << err);
+    if (current == messages.end())
         return false;
-    }
 
-    // XXX: very fishy place with const casting.
+    /// XXX: very fishy place with const-casting.
     auto new_position = reinterpret_cast<char *>(const_cast<unsigned char *>(current->get_payload().get_data()));
+    if (buffer().begin() == new_position)
+        return false;
```

Comment:
why?

Reply:
It's a replacement for old `stalled` logic: if we call next without resetting, then we've reached the end of message and should return `false`.

Reply:
IMHO it's harder to follow the logic than with simple flag. 
If you will leave it like that - please add the comment. 

BTW - does that casting cost nothing?

Reply:
`const_cast` is a compile time operation.